File: Processor/LocalGetter.cpp
#include "LocalGetter.h"

LocalGetter::~LocalGetter() 
{
    cleanupCache();
}

void LocalGetter::getFile() 
{
    const std::string directoryPath = "../BackendAndFrontend/Files";
    for (const auto &entry : std::filesystem::directory_iterator(directoryPath)) 
    {
        if (entry.is_regular_file()) 
        {
            cachedFilePath = entry.path().string();
            fileStream.open(cachedFilePath, std::ios::in);
            break;  // Only get one file
        }
    }
}

std::string LocalGetter::getFileName() 
{
    // Check if cachedFilePath is not empty
    if (cachedFilePath.empty()) 
    {
        std::cerr << "Error: File path is empty!" << std::endl;
        return;
    }

    // Find the position of the last directory separator
    size_t lastSeparator = cachedFilePath.find_last_of("/\\");

    // Find the position of the last period
    size_t lastPeriod = cachedFilePath.rfind('.');

    if (lastPeriod == std::string::npos || (lastSeparator != std::string::npos && lastPeriod < lastSeparator))
    {
        std::cerr << "Error: File extension not found or incorrect!" << std::endl;
        return;
    }

    // Extract the file name
    std::string fileName;
    if (lastSeparator != std::string::npos) 
    {
        fileName = cachedFilePath.substr(lastSeparator + 1, lastPeriod - lastSeparator - 1);
    }
    else 
    {
        fileName = cachedFilePath.substr(0, lastPeriod);
    }

    // Output the file name
    return fileName;
}


std::ifstream& LocalGetter::retrieveFile() 
{
    return fileStream;
}

void LocalGetter::removeFile() 
{
    if (!cachedFilePath.empty()) 
    {
        fileStream.close();  // Close the stream before deleting the file
        std::filesystem::remove(cachedFilePath);
        cachedFilePath.clear();
    }
}

void LocalGetter::cleanupCache() 
{
    if (fileStream.is_open()) 
    {
        fileStream.close();
    }
    cachedFilePath.clear();
}

File: Processor/LocalGetter.go
package filegetter

import (
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"
)

type LocalGetter struct {
    cachedFilePath string
}

// NewLocalGetter initializes and returns a new LocalGetter instance
func NewLocalGetter() *LocalGetter {
    return &LocalGetter{}
}

// GetFile fetches a file from a specified directory and caches its path.
func (lg *LocalGetter) GetFile() error {
    directoryPath := "../BackendAndFrontend/Files"

    files, err := ioutil.ReadDir(directoryPath)
    if err != nil {
        return err
    }

    for _, file := range files {
        if !file.IsDir() {
            lg.cachedFilePath = filepath.Join(directoryPath, file.Name())
            return nil
        }
    }
    return fmt.Errorf("no regular files found in directory: %s", directoryPath)
}

// GetFileName extracts and returns the name of the cached file (without the extension).
func (lg *LocalGetter) GetFileName() (string, error) {
    if lg.cachedFilePath == "" {
        return "", fmt.Errorf("file path is empty")
    }
    return strings.TrimSuffix(filepath.Base(lg.cachedFilePath), filepath.Ext(lg.cachedFilePath)), nil
}

// RetrieveFile returns the content of the cached file.
func (lg *LocalGetter) RetrieveFile() ([]byte, error) {
    return ioutil.ReadFile(lg.cachedFilePath)
}

// RemoveFile removes the cached file from the filesystem.
func (lg *LocalGetter) RemoveFile() error {
    if err := os.Remove(lg.cachedFilePath); err != nil {
        return err
    }
    lg.cachedFilePath = ""
    return nil
}

// CleanupCache cleans up the cached file path.
func (lg *LocalGetter) CleanupCache() {
    lg.cachedFilePath = ""
}
